### ISSUE TITLE
E712: is based on a misreading of PEP8

### ISSUE THREAD

**Issue by glyph** [PyCQA/pycodestyle#696]
Quoth the specification,

- Don't compare boolean values to True or False using ==.

  ```
  Yes:   if greeting:
  No:    if greeting == True:
  Worse: if greeting is True:
  ```

pycodestyle considers both the "yes" and the "worse" case equivalently good; if anything, it should complain *more* about the "Worse" case than the "Yes" one.

It also complains, I believe spuriously, about SQLAlchemy's overridden use of `==` [as described in this Stack Overflow question](https://stackoverflow.com/questions/18998010/flake8-complains-on-boolean-comparison-in-filter-clause).

Could this be restricted to just emitting the warning in the literal case of `if ... == True:`, or `is True:` which is probably redundant, and leaving more complex expressions alone?

**Comment by sigmavirus24** [comment]
I suspect that the `Worse` clause was added after the check was written. (That check has existed for years.)

I believe the specification doesn't scope this to just `if` statements. The examples use `if` statements but they intend for it to apply regardless of where it occurs. I agree it's a pain in the neck in this case and it could be improved to also flag `greeting is True`. (That may have regressed as I thought it used to flag that.)

**Comment by EndlessDex** [comment]
I would like to take on this issue; however, I am not sure about how to deal with the `not` cases. Below is my current behavior. Should `is not True`  and related be allowed?

```
if x is True:
if x is False:
if x is None: Good

if True is x:
if False is x:
if None is x: Good

if x is not True:
if x is not False:
if x is not None: Good

if True is not x:
if False is not x:
if None is not x: Good
```

**Comment by lordmauve** [comment]
We had a lint warning on `if x is True` in `$large_organisation`, and I saw it cause people to break code by rewriting correct code like

```python
if x is True:
    do_something()
elif x > 0:
   do_something_else()
```

to

```python
if x:
   do_something()
elif x > 0:
   do_something_else()
```

The case PEP-8 says is "Worse" above has different semantics, and pycodestyle shouldn't recommend changes that make a semantic difference to code.

**Comment by akern40** [comment]
I'm not sure what the conversation is around this issue, but I've just run across the same question. It seems that pylint does not recommend the replacement `if x is True`, only `if x`. Was the consensus from @lordmauve's response above that a change in pycodestyle's recommendation would be a poor decision? I'd be happy to try to close this issue with a PR for either the code (to adhere to PEP8) or to the documentation (to explain why there's a contradiction to PEP8).

**Comment by FichteFoll** [comment]
`if x is True:` has valid use cases, notably when `x` is not already known to be a boolean because only then the alternative `if x:` is actually equivalent. As such and because of what @lordmauve said with an actual example from people misinterpreting the lint error, it should not be changed imho.

Mentioning this difference to PEP8 in documentation with an explanation sounds good to me, though.

**Comment by akern40** [comment]
I will say that (although I'll happily defer to a group consensus) I do think that the lint error should be changed. In particular, if the goal is to ensure that pycodestyle does not recommend semantically different code then the error already needs to be changed. Perhaps a middle ground is to change some of the language of the current error without changing the logic. I believe the current error on `if cond == True:` is
```
comparison to True should be 'if cond is True:' or 'if cond:'
```
but one alternative is
```
comparison to True should be 'if cond:', or 'if cond is True:' for explicit comparison to boolean
```
or something of that sort? This gives slight priority to `if cond:` (the semantically identical syntax) and tells users why there are two different suggestions (and when to use them). This still has a "contradiction" to PEP8, but recognizes that PEP8 does not distinguish between the semantic differences of the three syntaxes. This could additionally be accompanied by a note in the documentation clarifying the lint error. Thoughts?

**Comment by FichteFoll** [comment]
SGTM and I'm in favor of having sufficient documentation inside the lint as opposed to some website.

**Comment by sigmavirus24** [comment]
> I'm not sure what the conversation is around this issue, but I've just run across the same question. It seems that pylint does not recommend the replacement `if x is True`, only `if x`. Was the consensus from @lordmauve's response above that a change in pycodestyle's recommendation would be a poor decision? I'd be happy to try to close this issue with a PR for either the code (to adhere to PEP8) or to the documentation (to explain why there's a contradiction to PEP8).

Does pylint advise against the check with `is True`? If not I don't see the problem you're trying to solve.

I also don't believe we need to change the lint at this point.

**Comment by akern40** [comment]
> Does pylint advise against the check with is True? If not I don't see the problem you're trying to solve.

Sorry, the pylint conversation was just trying to give a comparison to another linter's solution to the PEP8 weirdness.

The problem I was trying to address is providing clarity between the apparent contradiction between PEP8 and pycodestyle.  PEP8 (without semantic distinction) lays out three "tiers" of preference, and pydocstyle (without semantic distinction) suggests "Tier 3" then "Tier 1".

> `comparison to True should be 'if cond:', or 'if cond is True:' for explicit comparison to boolean`

This suggestion attempts to help users by providing PEP8's "Tier 1" as its first suggestion, and only providing "Tier 3" with  context for when to use it.

> I also don't believe we need to change the lint at this point.

If that's the consensus, should this issue be closed?

**Comment by Proofer** [comment]
I encountered an E712 warning from PyCharm yesterday in a case in which I think there is no coding alternative:

```
BAD_C_ENDS = '-WS', '-U', '-R', '-P'
# df is a pandas dataframe; the code selects rows that meet two criteria and don't meet a third, and copies them:
df = df[
        df['A'] == 'foo' & 
        df['B'] == 'bar' & 
        df['C'].str.endswith(BAD_C_ENDS) == False
       ] 
# PEP 8: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
```
Using `not` instead of `== False` causes a runtime `ValueError: The truth value of a Series is ambiguous.`

As a not-very-experienced Python user it seems to me that I should not get a style warning where I have no alternative other than doing an additional `drop` (i.e., delete) operation. On the other hand, I can turn off display of the error.

**Comment by asottile** [comment]
@Proofer yes pandas (and sqlalchemy, and numpy, and a few others) have odd overridden `__eq__` where this is the only way you can write this.  it looks wrong in any other context -- you should use a `# noqa: E712` comment in such cases to ignore the linter for the special case

**Comment by daviewales** [comment]
@proofer, there is a way to rewrite your code to avoid comparison to `False` by using the pandas `~` operator, which negates a boolean Series:

``` python
df = df[
        (df['A'] == 'foo') & 
        (df['B'] == 'bar') & 
        (~df['C'].str.endswith(BAD_C_ENDS))
       ] 
```

**Comment by asottile** [comment]
@pedrowaksman that snippet does not have an error -- are you confused with something else?

**Comment by asottile** [comment]
as is normal with numpy you'll want to use their types and conventions.  numpy has some pretty unconventional bits such as their own bool types and == projections

**Comment by asottile** [comment]
I'm going to close this out.  it definitely isn't changing and you're free to disable the code if it's causing problems in unconventional uses
